### PR TITLE
Create Certora spec for Reserve spender

### LINKED ISSUE #8513: Write Reserve certora spec protecting spender and transfer functions
### Expected Behavior

Please describe the behavior you are expecting

### Current Behavior

No Reserve spec

### PR DESCRIPTION
**Pull request by m-chrzan**
### Description

This PR adds a Certora spec for the Reserve contract, so far with one rule: checking that the Reserve's funds cannot be moved by an address that's not an approved spender.

### Other changes

* Added this as a CI job
* Set all spec-related scripts as executable

### Tested

Spec passes when running locally

### Related issues

- Fixes #8513

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -810,7 +810,7 @@ jobs:
     working_directory: ~/app
     docker:
       - image: circleci/openjdk:11-jdk-node
-    parallelism: 4
+    parallelism: 5
```
Comment (yorhodes) [inline .circleci/config.yml:813]: 🚀

File: packages/protocol/specs/harnesses/ReserveHarness.sol [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+contract ReserveHarness is Reserve {
+  constructor() public Reserve(true) {}
+
+  function contractBalance() public view returns (uint256) {
+    return address(this).balance;
+  }
+
+  function getExchangeAddress() public view returns (address) {
+    return registry.getAddressForOrDie(EXCHANGE_REGISTRY_ID);
```
Comment (yorhodes) [inline packages/protocol/specs/harnesses/ReserveHarness.sol]: super minor
```suggestion
    return getExchange();
```

File: packages/protocol/specs/reserve.spec [hunk 3]
```diff
@@ -0,0 +1,27 @@
+pragma specify 0.1
+
+methods {
+  isSpender(address) returns bool envfree
+  isExchangeSpender(address) returns bool envfree
+  getExchangeAddress() returns address envfree
+  contractBalance() returns uint256 envfree
+}
+
+rule nonspenderCantReduceBalance(address a, method f) {
```
Comment (yorhodes) [inline packages/protocol/specs/reserve.spec]: nice!

File: .circleci/config.yml [hunk 4]
```diff
@@ -867,6 +867,15 @@ jobs:
               ./specs/scripts/applyHarness.sh
               ./specs/scripts/governance.sh
             fi
+      - run:
+          name: Build and run - Reserve
+          command: |
+            if [ ${CIRCLE_NODE_INDEX} == "4" ];
+            then
+              cd packages/protocol
+              chmod +x specs/scripts/*.sh
```
Comment (yorhodes) [inline .circleci/config.yml]: did some quick googling and looks like git tracks executable bit of file permissions?

File: packages/protocol/specs/reserve.spec [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  require(e.msg.sender == a);
+  calldataarg arg;
+  sinvoke f(e, arg);
+
+  uint256 balanceAfter = invoke contractBalance();
+  if (f.isFallback) {
+    assert balanceAfter >= balanceBefore;
+  } else {
+    assert balanceAfter == balanceBefore;
+  }
```
Comment (yorhodes) [inline packages/protocol/specs/reserve.spec:28]: this makes sense but is it any more useful than 

```suggestion
  assert balanceAfter >= balanceBefore;
```

**Review by yorhodes — APPROVED** [review]
excited to build these out

File: packages/protocol/specs/reserve.spec (hunk 5 above)
Comment (m-chrzan) [reply packages/protocol/specs/reserve.spec:28]: I guess just being as precise as possible with the specification.

File: .circleci/config.yml (hunk 4 above)
Comment (m-chrzan) [reply .circleci/config.yml]: Cool, I copied this over from the other configs, I'll clean all of them up instead to not need the `chmod` here.

File: packages/protocol/specs/reserve.spec [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  isExchangeSpender(address) returns bool envfree
+  getExchangeAddress() returns address envfree
+  contractBalance() returns uint256 envfree
+}
+
+rule nonspenderCantReduceBalance(address a, method f) {
+  require(!isSpender(a));
+  require(!isExchangeSpender(a));
+  require(getExchangeAddress() != a);
+  uint256 balanceBefore = invoke contractBalance();
```
Comment (shellygr) [inline packages/protocol/specs/reserve.spec]: you could omit the `invoke` here - if it reverts, value of `balanceBefore` is undefined.

File: packages/protocol/specs/reserve.spec [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  isSpender(address) returns bool envfree
+  isExchangeSpender(address) returns bool envfree
+  getExchangeAddress() returns address envfree
+  contractBalance() returns uint256 envfree
+}
+
+rule nonspenderCantReduceBalance(address a, method f) {
+  require(!isSpender(a));
+  require(!isExchangeSpender(a));
+  require(getExchangeAddress() != a);
```
Comment (shellygr) [inline packages/protocol/specs/reserve.spec]: are all of these requirements necessary? Do we know it's safe to assume? In particular the 3rd one

File: packages/protocol/specs/reserve.spec [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  uint256 balanceBefore = invoke contractBalance();
+
+  env e;
+  require(e.msg.sender == a);
+  calldataarg arg;
+  sinvoke f(e, arg);
+
+  uint256 balanceAfter = invoke contractBalance();
```
Comment (shellygr) [inline packages/protocol/specs/reserve.spec]: ditto remove `invoke`

File: packages/protocol/specs/reserve.spec [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+rule nonspenderCantReduceBalance(address a, method f) {
+  require(!isSpender(a));
+  require(!isExchangeSpender(a));
+  require(getExchangeAddress() != a);
+  uint256 balanceBefore = invoke contractBalance();
+
+  env e;
+  require(e.msg.sender == a);
+  calldataarg arg;
+  sinvoke f(e, arg);
```
Comment (shellygr) [inline packages/protocol/specs/reserve.spec]: no need for `sinvoke`, it's the default for solidity function applications

File: packages/protocol/specs/reserve.spec (hunk 3 above)
Comment (shellygr) [inline packages/protocol/specs/reserve.spec]: what is `a` here? maybe you can give more descriptive names

File: packages/protocol/specs/scripts/reserve.sh [hunk 10]
```diff
@@ -0,0 +1,7 @@
+certoraRun contracts/common/Registry.sol specs/harnesses/ReserveHarness.sol \
+  --link ReserveHarness:registry=Registry \
+  --verify ReserveHarness:specs/reserve.spec \
+  --optimistic_loop \
+  --settings -ciMode=true \
```
Comment (shellygr) [inline packages/protocol/specs/scripts/reserve.sh]: you could do instead `--short_output`

**Review by shellygr — COMMENTED** [review]
Gave a few comments, but the most important thing is did you try to plant a bug in the code and see if the spec catches it?

File: packages/protocol/specs/reserve.spec (hunk 7 above)
Comment (m-chrzan) [reply packages/protocol/specs/reserve.spec]: Yes, spenders from the `isSpender` mapping and core stable token Exchanges are the exceptional addresses that can spend Reserve's balance. Thanks for pointing this out though, the last two can be simplified into a single call to `isAllowedToSpendExchange`.

Comment (m-chrzan) [reply packages/protocol/specs/reserve.spec]: Ah, actually `isAllowedToSpendExchange` is a modifier so the code can't be simplified that way.

**Pull request merged**